src/net_processing.cpp

```diff
@@ -1723,6 +1730,12 @@ void PeerManagerImpl::BlockConnected(const std::shared_ptr<const CBlock>& pblock
             m_txrequest.ForgetTxHash(ptx->GetWitnessHash());
         }
     }
+    auto stalling_timeout = m_block_stalling_timeout.load();
+    // In case the dynamic timeout was doubled once or more, reduce it slowly back to its default value
+    const auto new_timeout = std::max(std::chrono::duration_cast<std::chrono::seconds> (stalling_timeout * 0.85), DEFAULT_BLOCK_STALLING_TIMEOUT);
+    if (m_block_stalling_timeout.compare_exchange_strong(stalling_timeout, new_timeout) && stalling_timeout != new_timeout) {
```

Comment:
Let's say `compare_exchange_strong` triggers the `false` path, meaning `m_block_stalling_timeout` was updated elsewhere.

If I understand this method correctly, it will update `stalling_timeout` to the new value of `m_block_stalling_timeout`. Then, assuming `stalling_timeout != new_timeout` passes, the logging will be wrong?
1. We haven't decreased the actual value *here*
2. The latest value is *for sure* something different than `new_timeout`

I suspect this event could be even impossible in our execution, but then I'm wondering if there's a way to express this observation in code, rather than having a meaningless code branch.

Reply:
According to https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange, `compare_exchange_strong` returns `true` if the underlying atomic value was successfully changed, `false` otherwise.
So my understanding is that if nothing was updated because `stalling_timeout` was changed by some other thread, the logging statement won't be executed.

Reply:
>According to https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange, compare_exchange_strong returns true if the underlying atomic value was successfully changed, false otherwise.

I think this is only a part of what's happening, and if you consider the latter (also from your link), my example would make more sense?

```
Atomically compares the [object representation](https://en.cppreference.com/w/cpp/language/object) (until C++20)[value representation](https://en.cppreference.com/w/cpp/language/object) (since C++20) of *this with that of expected, and if those are bitwise-equal, replaces the former with desired (performs read-modify-write operation). Otherwise, loads the actual value stored in *this into expected (performs load operation).
```

Reply:
I thought you meant `stalling_timeout` possibly being updated, which might have had an influence on the second condition (`stalling_timeout != new_timeout`).
My point was that this couldn't possibly matter here because if `stalling_timeout` is updated, that must go together with `compare_exchange_strong()` returning false, and the two conditions are combined with an `&&`-operator so the entire clause is false (and the second condition will not even be evaluated due to short-circuiting).

Note that I changed the order of the two condition with the last push, so `stalling_timeout != new_timeout` is now the first of the two.

Reply:
You are right, I was wrong about the previous version. The updated version is better anyway :) feel free to resolve this.